include/IVideoDriver.h

```diff
@@ -191,6 +192,7 @@ namespace video
 		virtual void beginQuery(IQueryObject* query, const size_t& index) = 0;
```

Comment:
you can nuke `E_4X3_TRANSFORMATION_STATE` and `E_PROJECTION_TRANSFORMATION_STATE`

and just do a built-in include with the default definition of a per-camera and per-scene node descriptor set (UBOs only really)

Reply:
you can also kill all `draw2D*` functions

Reply:
@Crisspl still need to nuke the transformation state stuff (just provide a built-in GLSL include with a default descriptor set)

Reply:
i dont see how "built-in GLSL include with a default descriptor set" and "transformation state stuff" are substitutes?

Reply:
First replace everything with `matrix***SIMD` types, then

# For people and parts of the engine only using the `video` namespace

`whatever/path/we/have/for/builtins/DefaultCameraData.glsl`
```glsl
struct DefaultCameraData
{
    mat4x3 view;
    mat4x3 viewInverse;
    mat4 proj;
    mat4 projInverse;
    mat4 projView;
    mat4 projViewInverse;
};
```
`whatever/path/we/have/for/builtins/DefaultSceneNodeData.glsl`
```glsl
struct DefaultSceneNodeData
{
    mat4x3 view;
    mat4x3 viewInverse;
    mat4 proj;
    mat4 projInverse;
    mat4 projView;
    mat4 projViewInverse;
};
```

`whatever/path/we/have/for/builtins/defaultCameraUBO/location/${loc}.glsl`
```glsl
#include "whatever/path/we/have/for/builtins/DefaultCameraData.glsl"
layout(set=1, location=${loc}, row_major) DefaultPerViewUBO
{
    DefaultCameraData data;
} defaultPerViewUBO;
```
`whatever/path/we/have/for/builtins/defaultSceneNodeUBO/location/${loc}.glsl`
```glsl
#include "whatever/path/we/have/for/builtins/DefaultSceneNodeData.glsl"
layout(set=1, location=${loc}, row_major) DefaultPerSceneNodeUBO
{
    DefaultSceneNodeData data[];
} defaultPerSceneNodeUBO;
```

Then people will need to know to push new data (via streaming update or upload everything up-front then rebind at different offset before every draw).

# Fixing the `scene` namespace

Will be piss-easy. We allow each scene node to hold a descriptor set + offsets, and also the scene manager (the global desc set).

```cpp
class ISceneNode
{
...
setDescriptorSet(...) {}
getDescriptorSet(...) {}
setDynamicOffsets(...) {}
getDynamicOffsets(...) {}
...
IGPUDescriptorSet* descSet;
core::refctd_dynamic_array<uint32_t> dynamicOffsets;
...
}
```

Now the difference is that during the override of `ISceneNode::render()` the camera scene node will bind its descriptor set to slot 1, but a regular scene node will bind to slot 2.

Any meshbufferscenenode will take the descriptor set + dynamic offsets attached to a meshbuffer and bind it to slot 3, as well as push the constants just before the draw.

And obviously `ISceneManager::drawAll()` shall begin by binding descriptor set to slot 0.


## How the data will be updated

We will need a `scene::DescriptorSetUpdater` class and a default instantiation of it that will run after `ISceneNode::OnAnimate()` is finished, think of it like an animation call-back.

I see a method like `scene::DescriptorSetUpdater::update(scene::ISceneNode*,video::IGPUDescriptorSet*,core::smart_refctd_dynamic_array<uint32_t>* offsets)`

The default one for cameras will stream update the buffer in location 0 at offset assigned with the matrices as expected by the `DefaultPerViewUBO`.

The default one for mesh scene nodes will do the same for `DefaultPerSceneNodeUBO`.

User can derive and make a different descriptor-set-updater.

I plan to replace this prototype with something much faster when I come to do #96 and bucket-rendering.

Reply:
@Crisspl can you back these comments up somewhere, I'm scared I'll have to go over it again in the near future when you fix `irr::scene`

@AnastaZIuk read this thread, relevaant to the loader refactor.